### PR TITLE
Restore ExecuteInDefaultAppDomain implementation

### LINKED ISSUE #20131: New implementation for ICLRRuntimeHost::ExecuteInDefaultAppDomain  for .NetCore 3.0
XAML Visual Diagnostics and Edit-N-Continue need a mechanism for bootstrapping Microsoft.VisualStudio.DesignTools.WpfTap.dll into a running application for Visual Diagnostics to work

Here is how I think the injection may work in .NET Core 3.0:

1.	Find coreclr.dll instance that you like. It can be as simple as calling GetModuleHandle(“coreclr.dll”). Having multiple coreclr instances in the one process is not really a scenario today, and designing this for it would make us rathole and never get done. If multiple coreclr instances in a single process become a mainstream scenario, this code will need to change.
2.	Call coreclr!GetCLRRuntimeHost (existing API in coreclr.dll) to get ICLRRuntimeHost interface
3.	Call ICLRRuntimeHost::ExecuteInDefaultAppDomain like you are doing today. This method exists in CoreCLR, but it returns E_NOTIMPL. This would be work item for CoreCLR to get implementation of this method filled in.

### PR DESCRIPTION
**Pull request by sdmaclea**
Restore deleted ExecuteInDefaultAppDomain implentation
Enable for CoreCLR
Cleanup for CoreCLR

@jkotas @vitek-karas @jeffschwMSFT 

Fixes #20131

### CODE REVIEW

**Comment by sdmaclea** [comment]
This is basically the same as the code which had been removed from CoreCLR earlier
https://github.com/dotnet/coreclr/blob/cb3773542726fda1267e81596bd37654f418cff4/src/vm/corhost.cpp

There are a few minor changes for formatting, simplification, and robustness.  I'll comment on them

File: src/vm/corhost.cpp [hunk 1]
```diff
@@ -520,9 +520,85 @@ HRESULT CorHost2::ExecuteInDefaultAppDomain(LPCWSTR pwzAssemblyPath,
...
+        pThread = SetupThreadNoThrow(&hr);
+        if (pThread == NULL)
+        {
+            goto ErrExit;
+        }
+    }
+
+    _ASSERTE (!pThread->PreemptiveGCDisabled());
+
+    _ASSERTE (SystemDomain::GetCurrentDomain()->GetId().m_dwId == DefaultADID);
```
Comment (sdmaclea) [inline src/vm/corhost.cpp]: Removed code to enter and exit the default domain and replaced with this assert.

**Review by jkotas — APPROVED** [review]

File: src/vm/corhost.cpp [hunk 2]
```diff
@@ -520,9 +520,85 @@ HRESULT CorHost2::ExecuteInDefaultAppDomain(LPCWSTR pwzAssemblyPath,
...
+
+    INSTALL_UNHANDLED_MANAGED_EXCEPTION_TRAP;
+    INSTALL_UNWIND_AND_CONTINUE_HANDLER;
```
Comment (sdmaclea) [inline src/vm/corhost.cpp]: Moved the INSTALL* UNINSTALL* out of the try block.
Added the INSTALL_UNHANDLED_MANAGED_EXCEPTION_TRAP and its uninstall based on ExecuteAssembly code above.

**Comment by jeffschwMSFT** [comment]
Thanks @sdmaclea.

**Comment by sdmaclea** [comment]
I haven't tested this yet.  Since it is copy and paste it may not need much real testing.  Also since it is implementing an unimplemented feature it is unlikely to break anything.

I would like to add unit tests somewhere.  This is a bit of a difficult scenario to setup and test.  Right now it seems the easiest place would be in the core-setup repo, but will have to wait until these bits percolate to that repo.  

Would like to test these scenarios
* Assembly not found returns reasonable HRESULT
* Type not found returns reasonable HRESULT
* Method signature incorrect returns reasonable HRESULT
* Simple Method works correctly
* Method with an uncaught managed exception return a reasonable HRESULT
* New C++ thread
* Preexisting .NET thread

Probably the simplest setup would be to P/Invoke into native code (dllmap would make this easier) and then have the native code test the ExecuteInAppDomain scenario.

**Comment by jeffschwMSFT** [comment]
In general I do not believe we have much direct validation for our hosting surface area.  We tend to use corerun as both the test runner and test validation.  I do think it would be worth adding a simple validation test, but I do not think it warrants an extensive validation.

**Comment by sdmaclea** [comment]
test  Windows_NT arm64 Cross Checked Innerloop Build and Test
test Windows_NT arm Cross Checked Innerloop Build and Test

**Comment by sdmaclea** [comment]
@dotnet-bot help

**Comment by sdmaclea** [comment]
@mmitche @weshaggard Looks like the request for help retriggered all the jobs...  Maybe because the dotnet bot comments contains triggers???

**Comment by mmitche** [comment]
@sdmaclea Nope, it ignores the bot comments.  It did so for some other reason.

**Comment by vitek-karas** [comment]
I think the tests should be in the coreclr repo. This has basically nothing to do with the hosting code in core-setup. Also coreclr already has tests which mix managed and native code. core-setup might be a bit trickier to setup that way.
Take a look at for example this [interop test](https://github.com/dotnet/coreclr/blob/master/tests/src/Interop/MarshalAPI/FunctionPointer/GetFcnPtrForDel_Negative_Security.cs) which uses a native component.

**Comment by sdmaclea** [comment]
I added a unit test which is a fair approximation of WPF use case @chabiss described in #20131.

To get the tests to work I needed to add support for `IID_ICLRRuntimeHost` and the call sequence needed to include `host->Start()` and `host->Stop()`

The unit test starts from C#, calls out to a DLL, which find the runtime dll, gets the CLRRuntimeHost interface, call `host->Start()`, `host->ExecuteInDefaultAppDomain()`, and `host->Stop()`

The tests are reasonably exhaustive, however 
* It does not test loading a separate assembly from a path
* It does not test from a new C++ thread
* It does not test process attach.

@chabiss, @vitek-karas, @jkotas PTAL

**Comment by sdmaclea** [comment]
Apparently mscoree.h is not being found on Linux.  I will take a look in the morning.

**Review by vitek-karas — APPROVED** [review]
Please do add a test which loads a new assembly through this, but otherwise LGTM.

File: src/vm/corhost.cpp [hunk 3]
```diff
@@ -1272,6 +1348,10 @@ HRESULT CorHost2::QueryInterface(REFIID riid, void **ppUnk)
     // GetConfiguration to obtain that interface.
     if (riid == IID_IUnknown)
         *ppUnk = static_cast<IUnknown *>(static_cast<ICLRRuntimeHost *>(this));
```
[the remark this answers is not in the document]
Comment (sdmaclea) [reply src/vm/corhost.cpp]: Done

**Comment by sdmaclea** [comment]
I was hoping the prebuilt mscoree.h would work easily on *nix.  It turned out not so simple.  There were a bunch of compile errors, so since this feature is targeting WPF.  It seemed OK to disable for now.

When we add a native hosting SDK, it would be good to refactor this test and enable on all platforms.

For now I disabled the test on *nix.

**Comment by sdmaclea** [comment]
OK added an InjectedCode classlib test to test loading a separate assembly from a path.

**Comment by sdmaclea** [comment]
test Windows_NT arm64 Cross Checked Innerloop Build and Test

**Comment by sdmaclea** [comment]
test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test
test Windows_NT arm64 Cross Checked Innerloop Build and Test

**Pull request merged**